File: Algorithm/C++/435.non-overlapping-intervals.cpp
/*
 * @lc app=leetcode id=435 lang=cpp
 *
 * [435] Non-overlapping Intervals
 */

// @lc code=start

#include <vector>
#include <algorithm>

class Solution {
public:
    int eraseOverlapIntervals(std::vector<std::vector<int>>& intervals) {
        std::sort(intervals.begin(), intervals.end(), [](const std::vector<int>& v1, const std::vector<int>& v2) -> bool {
            if (v1[0] != v2[0]) {
                return v1[0] < v2[0];
            } else {
                return v1[1] < v2[1];
            }
        });
        

        int pre_end = intervals[0][1], result = 0;
        for (int i = 1; i < intervals.size(); i++) {
            if (intervals[i][0] < pre_end) {
                result++;
                if (intervals[i][1] < pre_end) {
                    pre_end = intervals[i][1];
                }
            } else {
                pre_end = intervals[i][1];
            }
        }
        return result;
    }
};
// @lc code=end

File: Algorithm/Python/435.non-overlapping-intervals.py
#
# @lc app=leetcode id=435 lang=python
#
# [435] Non-overlapping Intervals
#

# @lc code=start
class Solution(object):
    def eraseOverlapIntervals(self, intervals):
        """
        :type intervals: List[List[int]]
        :rtype: int
        """
        def comparator(x1, x2):
            if x1[0] == x2[0]:
                return x1[1] - x2[1]
            else:
                return x1[0] - x2[0]

        intervals.sort(cmp = comparator)

        start, end = intervals[0]
        result = 0
        for i in range(1, len(intervals)):
            s, e = intervals[i]
            if s < end:
                result += 1
                if e < end:
                    start, end = s, e
            else:
                start, end = s, e
        return result

# @lc code=end

